File: 1-cpp-vs-python/opencv_python/convolution.py
import cv2
import numpy as np
import time


def apply_convolution(image, kernel):
    rows, cols = image.shape
    output = np.zeros((rows, cols), dtype=np.uint8)

    for i in range(1, rows - 1):
        for j in range(1, cols - 1):
            pixel_sum = 0
            for ki in range(-1, 2):
                for kj in range(-1, 2):
                    pixel_sum += kernel[ki + 1][kj + 1] * image[i + ki, j + kj]
            output[i, j] = min(max(pixel_sum, 0), 255)

    return output


if __name__ == "__main__":
    image = cv2.imread("../../assets/lion-king.jpg", cv2.IMREAD_GRAYSCALE)
    kernel = np.array([[1, 0, -1], 
                       [1, 0, -1], 
                       [1, 0, -1]], dtype=np.int32)

    start_time = time.time()
    convolved_image = apply_convolution(image, kernel)
    end_time = time.time()

    print(f"Convolution Time (Python): {end_time - start_time:.6f} seconds")
    cv2.imwrite("output_python.jpg", convolved_image)

File: 1-cpp-vs-python/opencv_cpp/convolution.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <chrono>

using namespace std;
using namespace cv;

void applyConvolution(const Mat &image, Mat &output, const int kernel[3][3])
{
    for (int i = 1; i < image.rows - 1; ++i)
    {
        for (int j = 1; j < image.cols - 1; ++j)
        {
            int pixel_sum = 0;
            for (int ki = -1; ki <= 1; ++ki)
            {
                for (int kj = -1; kj <= 1; ++kj)
                {
                    pixel_sum += kernel[ki + 1][kj + 1] * image.at<uchar>(i + ki, j + kj);
                }
            }
            output.at<uchar>(i, j) = saturate_cast<uchar>(pixel_sum);
        }
    }
}

int main()
{
    Mat image = imread("../../assets/lion-king.jpg", IMREAD_GRAYSCALE);
    Mat output = Mat::zeros(image.size(), CV_8U);

    int kernel[3][3] = {
        {1, 0, -1},
        {1, 0, -1},
        {1, 0, -1}};

    auto start_time = chrono::high_resolution_clock::now();
    applyConvolution(image, output, kernel);
    auto end_time = chrono::high_resolution_clock::now();

    chrono::duration<double> elapsed_time = end_time - start_time;
    cout << "Convolution Time (C++): " << elapsed_time.count() << " seconds" << endl;
    imwrite("output_cpp.jpg", output);
    return 0;
}
